server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -617,7 +626,7 @@ public RestoreInProgress applyChanges(final RestoreInProgress oldRestore) {
                 for (RestoreInProgress.Entry entry : oldRestore.entries()) {
                     Snapshot snapshot = entry.snapshot();
                     Updates updates = shardChanges.get(snapshot);
-                    if (updates.shards.isEmpty() == false) {
+                    if (updates != null && updates.shards.isEmpty() == false) {
```

Comment:
Now having an update here doesn't necessarily mean an update for every snapshot so we need the `0` check.